osquery/process/processes_stats.cpp

```diff
@@ -0,0 +1,146 @@
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include "processes_stats.h"
+
+#include <iostream>
+#include <sstream>
+
+#include <osquery/logger/logger.h>
+#include <osquery/utils/system/time.h>
+
+namespace osquery {
+
+CLI_FLAG(bool,
+         enable_processes_stats,
+         false,
+         "Enables the service that collects processes stats");
+
+CLI_FLAG(uint32,
+         processes_stats_interval,
+         2,
+         "Seconds that the processes stats service waits between scans");
+
+CLI_FLAG(
+    bool,
+    processes_stats_worker_only,
+    false,
+    "Enables process stats collection only for the osquery worker process");
+
+CLI_FLAG(uint32,
+         processes_stats_cpu_peak_samples,
+         30,
+         "How many past cpu usage samples to use to calculate the cpu peak "
+         "usage");
+
+std::shared_ptr<ProcessesStats>& ProcessesStats::getInstance() {
+  static auto processes_stats = std::make_shared<ProcessesStats>();
+  return processes_stats;
+}
+
+std::optional<ProcessStatsSample> ProcessesStats::getProcessStats(
+    std::uint32_t pid) {
+  std::lock_guard<std::mutex> process_stats_lock_guard(process_stats_lock_);
+  auto proc_it = processes_stats_states_.find(pid);
+  if (proc_it == processes_stats_states_.end()) {
+    return std::nullopt;
+  }
+
+  return ProcessStatsSample{proc_it->second.cpu_peak_usage,
+                            proc_it->second.cpu_usages.empty()
+                                ? static_cast<std::uint16_t>(0)
+                                : proc_it->second.cpu_usages.back()};
+}
+
+void ProcessesStats::updateProcessStats(std::uint32_t pid,
+                                        std::uint64_t total_cpu_time,
+                                        std::uint32_t update_interval) {
+  std::lock_guard<std::mutex> process_stats_lock_guard(process_stats_lock_);
+  auto proc_it = processes_stats_states_.find(pid);
+
+  if (proc_it == processes_stats_states_.end()) {
+    processes_stats_states_.emplace(
+        pid, ProcessStatsState{total_cpu_time, current_generation_.load()});
+  } else {
+    auto& proc_stats = proc_it->second;
+
+    /* If this pid is older than one generation,
+       it means it's a pid that disappeared
+       before the old pid cleanup procedure had a chance to run,
+       and now has reappeared */
+    if (proc_stats.generation < current_generation_ &&
+        (current_generation_ - proc_stats.generation) > 1) {
```

Comment:
I'm having trouble following all the logic here

Reply:
This handles the case of pid reuse. If at the next sample the pid/process is still in the state/cache it can mean two things:
1. It's the same process as before
2. It's a new process with the same pid and the cleanup hasn't run yet to remove it from the state/cache; we can identify that by checking the generation.

Reply:
how does it detect the pid reuse?

Reply:
In hindsight maybe it's slightly naive (I was working with only the data I had extracted).

The idea is that on sample you see the process, you store it in the map with the `pid` as key and set its generation to lets say, 0

Then the process exits, and you get another sample where you see that the process cannot be found anymore; you don't update the generation (it would be 1 otherwise).

Now on a new sample, the cleanup process might not have run yet, the information of that process might still be in the map; lets say then that a new process runs with the same pid as the old one, and that the new sample sees it.
Since the current generation is 2, and the process state generation is 0, it means that there has been at least one time that we haven't seen it, so it means that this is a different process.

Obviously the precision of this depended on the sample time and when the cleanup occurred.

But using the process start time might indeed be much better; I was initially unsure about the precision, given the issues we have on retrieving something stable for the `processes` table, but in this case we are actually storing the result for all the time we need it, so it won't change between comparisons.

Reply:
@directionless did this get resolved?  https://github.com/osquery/osquery/pull/7597#issuecomment-1245827168